### PR TITLE
initial jsonrpc v1 support

### PR DESCRIPTION
**Pull request by debris**
@tomusdrw if you agree this design I will add some tests

### CODE REVIEW

File: core/src/io.rs [hunk 1]
```diff
@@ -17,13 +17,52 @@ fn write_response(response: Response) -> String {
...
+/// By default compatible only with jsonrpc v2
 #[derive(Default)]
 pub struct MetaIoHandler<T: Metadata> {
+	compatibility: Compatibility,
 	methods: HashMap<String, RemoteProcedure<T>>,
 }
 
 impl<T: Metadata> MetaIoHandler<T> {
+	/// Creates new `MetaIoHandler` without any metadata compatible with specified protocol version.
+	pub fn new(compatibility: Compatibility) -> Self {
```
Comment (tomusdrw) [inline core/src/io.rs]: I would leave `new()` as is and add new `with_compatibility()` method.

File: core/src/io.rs [hunk 2]
```diff
@@ -17,13 +17,52 @@ fn write_response(response: Response) -> String {
...
+///
+/// By default compatible only with jsonrpc v2
 #[derive(Default)]
 pub struct MetaIoHandler<T: Metadata> {
+	compatibility: Compatibility,
 	methods: HashMap<String, RemoteProcedure<T>>,
 }
 
 impl<T: Metadata> MetaIoHandler<T> {
+	/// Creates new `MetaIoHandler` without any metadata compatible with specified protocol version.
```
Comment (tomusdrw) [inline core/src/io.rs]: Comment is invalid (it has metadata)

File: core/src/io.rs [hunk 3]
```diff
@@ -121,10 +160,12 @@ impl<T: Metadata> MetaIoHandler<T> {
 				let params = method.params.unwrap_or(Params::None);
 				let id = method.id;
 				let jsonrpc = method.jsonrpc;
+				let valid_version = self.compatibility.is_version_valid(jsonrpc);
 
-				let result = match self.methods.get(&method.method) {
-					Some(&RemoteProcedure::Method(ref method)) => method.call(params, meta),
-					_ => futures::failed(Error::new(ErrorCode::MethodNotFound)).boxed(),
+				let result = match (valid_version, self.methods.get(&method.method)) {
+					(false, _) => futures::failed(Error::invalid_request()).boxed(),
```
Comment (tomusdrw) [inline core/src/io.rs]: Would add more descriptive error for this.

File: core/src/io.rs [hunk 4]
```diff
@@ -157,6 +202,11 @@ impl IoHandler {
 	pub fn new() -> Self {
 		IoHandler::default()
 	}
+
+	/// Creates new `IoHandler` without any metadata compatible with specified protocol version.
+	pub fn new_compatible_with(compatibility: Compatibility) -> Self {
```
Comment (tomusdrw) [inline core/src/io.rs]: IMHO `with_compatbility` would be enough

File: core/src/types/request.rs [hunk 5]
```diff
@@ -11,7 +11,7 @@ use super::{Id, Params, Version, Value};
 pub struct MethodCall {
 	/// A String specifying the version of the JSON-RPC protocol.
 	/// MUST be exactly "2.0".
```
Comment (tomusdrw) [inline core/src/types/request.rs]: Comment should be updated.

File: core/src/types/request.rs [hunk 6]
```diff
@@ -29,7 +29,7 @@ pub struct MethodCall {
 pub struct Notification {
 	/// A String specifying the version of the JSON-RPC protocol.
 	/// MUST be exactly "2.0".
```
Comment (tomusdrw) [inline core/src/types/request.rs]: Comment should be updated.

File: core/src/types/request.rs [hunk 7]
```diff
@@ -233,20 +233,22 @@ fn request_deserialize_batch() {
...
 		Call::Notification(Notification {
-			jsonrpc: Version::V2,
+			jsonrpc: Some(Version::V2),
 			method: "update".to_owned(),
 			params: Some(Params::Array(vec![Value::U64(1)]))
 		})
 	]))
 }
 
+// TODO: fix this test. it's valid now, since jsonrpc v1 does not require "version" field
```
Comment (tomusdrw) [inline core/src/types/request.rs]: Indeed. (Perhaps add some unexpected field instead)

File: core/src/types/response.rs [hunk 8]
```diff
@@ -56,7 +56,7 @@ impl Output {
 	pub fn invalid_request(id: Id) -> Self {
```
Comment (tomusdrw) [inline core/src/types/response.rs]: This should probably get `Option<Version>` as well.

File: core/src/types/response.rs [hunk 9]
```diff
@@ -115,7 +115,7 @@ impl From<Error> for Response {
 	fn from(error: Error) -> Self {
```
Comment (tomusdrw) [inline core/src/types/response.rs]: This implementation is no longer valid. Should be more explicity: `Response::from_error(error, jsonrpc)`

File: core/src/types/version.rs [hunk 10]
```diff
@@ -3,7 +3,7 @@ use serde::{Serialize, Serializer, Deserialize, Deserializer, Error};
 use serde::de::Visitor;
 
 /// Protocol Version
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Copy)]
 pub enum Version {
```
Comment (tomusdrw) [inline core/src/types/version.rs]: Shouldn't this contain other versions? (Does spec say if `{"jsonrpc":"1.0", "method"....}` is a valid request?)

File: core/src/io.rs (hunk 1 above)
Comment (debris) [inline core/src/io.rs]: there is no `new` currently, so I will add it ;)

**Pull request merged**